src/coreclr/jit/morph.cpp

```diff
@@ -11565,6 +11580,68 @@ GenTree* Compiler::fgOptimizeHWIntrinsic(GenTreeHWIntrinsic* node)
 
 #endif
 
+//------------------------------------------------------------------------
+// fgOptimizeDivision: Optimizes division operations.
+//
+// Arguments:
+//   tree - the unchecked GT_DIV/GT_UDIV tree to optimize.
+//
+// Return Value:
+//   The optimized tree that can have any shape.
+//
+GenTree* Compiler::fgOptimizeDivision(GenTreeOp* div)
+{
+    assert(div->OperIs(GT_DIV, GT_UDIV));
+
+    GenTree* op1 = div->gtGetOp1();
+    GenTree* op2 = div->gtGetOp2();
+
+    // Folds cascade of DIVs/UDIVs into one DIV/UDIV and a multiplication
+    // of root and child values
+    if (op1->OperIs(GT_UDIV, GT_DIV) && op2->IsIntegralConst() && op1->gtGetOp2()->IsIntegralConst())
+    {
+
+        GenTree* chOp1 = op1->gtGetOp1();
+        GenTree* chOp2 = op1->gtGetOp2();
+
+        IntegralRange boundRange = IntegralRange::ForNode(op2, this);
+        int64_t       upperBound = IntegralRange::SymbolicToRealValue(boundRange.GetUpperBound());
```

Comment:
it's not clear to me why IntegralRange is used here - you already have child_val and root_val, don't you?